pkg/controller/database/redshift/controller.go

```diff
@@ -0,0 +1,218 @@
+
+	"github.com/crossplane/provider-aws/apis/database/v1alpha1"
+	"github.com/crossplane/provider-aws/pkg/clients/redshift"
+	"github.com/crossplane/provider-aws/pkg/controller/utils"
+)
+
+const (
+	errUnexpectedObject = "managed resource is not a Redshift custom resource"
+	errKubeUpdateFailed = "cannot update Redshift cluster custom resource"
+	errCreateFailed     = "cannot create Redshift cluster"
+	errModifyFailed     = "cannot modify Redshift cluster"
+	errDeleteFailed     = "cannot delete Redshift cluster"
+	errDescribeFailed   = "cannot describe Redshift cluster"
+	errUpToDateFailed   = "cannot check whether object is up-to-date"
+)
+
+// SetupRedshift adds a controller that reconciles Redshift clusters.
+func SetupRedshift(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.RedshiftGroupKind)
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Redshift{}).
+		Complete(managed.NewReconciler(
+			mgr, resource.ManagedKind(v1alpha1.RedshiftGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), newClientFn: redshift.NewClient, awsConfigFn: utils.RetrieveAwsConfigFromProvider}),
+			managed.WithInitializers(managed.NewNameAsExternalName(mgr.GetClient())),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube        client.Client
+	newClientFn func(*aws.Config) redshift.Client
+	awsConfigFn func(context.Context, client.Reader, *corev1.ObjectReference) (*aws.Config, error)
+}
+
+func (conn *connector) Connect(ctx context.Context, mgd resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mgd.(*v1alpha1.Redshift)
+	if !ok {
+		return nil, errors.New(errUnexpectedObject)
+	}
+
+	awsconfig, err := conn.awsConfigFn(ctx, conn.kube, cr.Spec.ProviderReference)
+	if err != nil {
+		return nil, err
+	}
+
+	c := conn.newClientFn(awsconfig)
+
+	return &external{kube: conn.kube, client: c}, nil
+}
+
+type external struct {
+	kube   client.Client
+	client redshift.Client
+}
+
+func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha1.Redshift)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+
+	rsp, err := e.client.DescribeClustersRequest(&awsredshift.DescribeClustersInput{
+		ClusterIdentifier: aws.String(meta.GetExternalName(cr)),
+	}).Send(ctx)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(redshift.IsErrorNotFound, err), errDescribeFailed)
+	}
+
+	// Describe requests can be used with filters, which then returns a list.
+	// But we use an explicit identifier, so, if there is no error, there should
+	// be only 1 element in the list.
+	instance := rsp.Clusters[0]
```

Comment:
I know this has been done elsewhere, but I would prefer that we check the length of the cluster slice instead of assuming that the API will always behave correctly.

Reply:
I have replaced that with the following snippet:

```go
if len(rsp.Clusters) > 1 {
		return managed.ExternalObservation{}, errors.Wrap(errors.New(errMultipleCluster), errMultipleCluster)
	}
instance := rsp.Clusters[0]
```

Reply:
@arush-sal my primary concern was actually with making sure the slice is not `nil`. So maybe we should change that check to `if len(rsp.Clusters) != 1` and update the error message appropriately

Reply:
@hasheddan We would get an empty slice only if there isn't a cluster with the name. In which we will also receive an err which will be caught by the above line. But still, as you suggested. I'll make the respective changes.